extension/intents/nicknames/routineExecutor.js

```diff
@@ -0,0 +1,118 @@
+/* globals log */
+
+import * as intentRunner from "../../background/intentRunner.js";
+
+export let pausedRoutineExecutor = null;
+export class RoutineExecutor {
+  constructor(
+    routineName,
+    subcommands,
+    programCounter = 0,
+    states = null,
+    startLoopProgramCounter = 0
+  ) {
+    this.routineName = routineName;
+    this.subcommands = subcommands;
+    this.programCounter = programCounter;
+    this.states = states;
+    this.startLoopProgramCounter = startLoopProgramCounter;
+  }
+
+  mapState(subcommand, state) {
+    const copySubcommand = JSON.parse(JSON.stringify(subcommand));
+    for (const slot in subcommand.slots) {
+      if (subcommand.slots[slot] in state) {
+        copySubcommand.slots[slot] = state[subcommand.slots[slot]];
+      }
+    }
+    return copySubcommand;
+  }
+
+  async runSubcommand(subcommand) {
+    let hadError = false;
+    let errorMessage = null;
+    if (this.states !== null) {
+      subcommand = this.mapState(subcommand, this.states[0]);
+    } else {
+      subcommand = JSON.parse(JSON.stringify(subcommand));
+    }
+    subcommand.routineExecutor = this;
+    subcommand.onError = message => {
+      hadError = true;
+      errorMessage = message;
+    };
+
+    log.info(
+      "  Running subintent",
+      subcommand,
+      subcommand.name,
+      subcommand.slots,
+      this.programCounter
+    );
+    await intentRunner.runIntent(subcommand);
+    if (hadError) {
+      log.info(
+        `Routine ${this.routineName} encountered error ${errorMessage} while running ${subcommand.utterance}`
+      );
+      return true;
+    }
+
+    return this._stop === true;
+  }
+
+  startLoop(states) {
+    if (this.states !== null) {
+      const exc = new Error("Encountered nested loops. Not yet supported.");
+      exc.displayMessage = "Encountered nested loops. Not yet supported.";
+      throw exc;
+    }
+    this.states = states;
+    this.startLoopProgramCounter = this.programCounter;
+  }
+
+  endLoop() {
+    this.states.shift();
```

Comment:
I think both here and in the intent for ending loops there's nothing that will stop you from ending a loop when there's no active loop. So you'll probably get some kind of exception?